pkg/vcdclient/gateway.go

```diff
@@ -1049,3 +1061,109 @@ func (client *Client) IsNSXTBackedGateway() bool {
+		return nil, "", nil, fmt.Errorf("error when getting defined entity: [%v]", err)
+	}
+	statusEntry, ok := defEnt.Entity["status"]
+	if !ok {
+		return nil, "", nil, fmt.Errorf("could not find 'status' entry in defined entity")
+	}
+	statusMap, ok := statusEntry.(map[string]interface{})
+	if !ok {
+		return nil, "", nil, fmt.Errorf("unable to convert [%T] to map", statusEntry)
+	}
+	virtualIpInterfaces := statusMap["virtual_IPs"]
+	if virtualIpInterfaces == nil {
+		return make([]string, 0), etag, &defEnt, nil
+	}
+
+	virtualIpInterfacesSlice, ok := virtualIpInterfaces.([]interface{})
+	if !ok {
+		return nil, "", nil, fmt.Errorf("unable to convert [%T] to slice of interface", virtualIpInterfaces)
+	}
+	virtualIpStrs := make([]string, len(virtualIpInterfacesSlice))
+	for ind, ipInterface := range virtualIpInterfacesSlice {
+		currIp, ok := ipInterface.(string)
+		if !ok {
+			return nil, "", nil, fmt.Errorf("unable to convert [%T] to string", ipInterface)
+		}
+		virtualIpStrs[ind] = currIp
+	}
+	return virtualIpStrs, etag, &defEnt, nil
+}
+
+// This function will modify the passed in defEnt
+func (client *Client) updateRDEVirtualIps(ctx context.Context, updatedIps []string, etag string,
+	defEnt *swaggerClient.DefinedEntity) (*http.Response, error) {
+	statusEntry, ok := defEnt.Entity["status"]
+	if !ok {
+		return nil, fmt.Errorf("could not find 'status' entry in defined entity")
+	}
+	statusMap, ok := statusEntry.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("unable to convert [%T] to map", statusEntry)
+	}
+
+	statusMap["virtual_IPs"] = updatedIps
+	_, httpResponse, err := client.apiClient.DefinedEntityApi.UpdateDefinedEntity(ctx, *defEnt, etag, client.ClusterID)
+	if err != nil {
+		return httpResponse, fmt.Errorf("error when updating defined entity: [%v]", err)
+	}
+	return nil, nil
+}
+
+func (client *Client) addVirtualIpToRDE(ctx context.Context, addIp string) error {
+	for {
+		currIps, etag, defEnt, err := client.GetRDEVirtualIps(ctx)
+		if err != nil {
+			return fmt.Errorf("error for getting current vips: [%v]", err)
+		}
+		updatedIps := append(currIps, addIp)
+		httpResponse, err := client.updateRDEVirtualIps(ctx, updatedIps, etag, defEnt)
+		if err != nil {
+			if httpResponse.StatusCode == 412 {
+				continue
+			}
+			return fmt.Errorf("error when adding virtual ip to RDE: [%v]", err)
+		}
+		break
+	}
+	return nil
+}
+
+func (client *Client) removeVirtualIpFromRDE(ctx context.Context, removeIp string) error {
+	for {
+		currIps, etag, defEnt, err := client.GetRDEVirtualIps(ctx)
+		if err != nil {
+			return fmt.Errorf("error for getting current vips: [%v]", err)
+		}
+		// currIps is guaranteed not to be nil by GetRDEVirtualIps
+		if len(currIps) == 0 {
+			return fmt.Errorf("no RDE virtual ips to remove")
+		}
+
+		// form updated virtual ip list
+		updatedIps := make([]string, len(currIps)-1)
+		updatedInd := 0
+		for _, ip := range currIps {
+			if ip == removeIp {
+				continue // for inner loop
```

Comment:
Here we can set a boolean to ensure that if the ip is already removed, a retry will be a NOP. Since we have an unending loop, it will be useful to add safeguards here.

Reply:
This inner loop will terminate because it only iterates through the current ip's. It needs to complete this loop to add the ips that will be used.

The outer loop is unending until the request does not have a 412 code